marlowe-runtime/sync/Language/Marlowe/Runtime/Sync/Database/PostgreSQL/GetNextSteps.hs

```diff
@@ -0,0 +1,430 @@
+      rollbackBlock.id :: bytea?,
+      rollbackBlock.blockNo :: bigint?
+    FROM marlowe.block
+    LEFT JOIN marlowe.block AS rollbackBlock ON block.rollbackToBlock = rollbackBlock.id
+    WHERE block.id = $1 :: bytea
+  |]
+  where
+    decodeResult Nothing = RolledBack Genesis
+    decodeResult (Just (Just slot, Just hash, Just block)) = RolledBack $ At BlockHeader
+      { slotNo = fromIntegral slot
+      , headerHash = BlockHeaderHash hash
+      , blockNo = fromIntegral block
+      }
+    decodeResult _ = BeforeTip
+
+data NextTxIds = NextTxIds
+  { nextBlock :: BlockHeader
+  , nextApplyTxIds :: [TxId]
+  , nextWithdrawalTxIds :: [TxId]
+  }
+
+getNextTxIds :: ContractId -> ChainPoint -> T.Transaction (Maybe NextTxIds)
+getNextTxIds (ContractId TxOutRef{..}) point = T.statement params $ fmap (NextTxIds <$> decodeBlockHeader <*> decodeApplyTxIds <*> decodeWithdrawalTxIds) <$>
+  [maybeStatement|
+    WITH params (createTxId, createTxIx, afterSlot) AS
+      ( SELECT $1 :: bytea, $2 :: smallint, $3 :: bigint
+      )
+    , nextApplyTxIds (slotNo, blockHeaderHash, blockNo, txIds) AS
+      ( SELECT
+          block.slotNo,
+          (ARRAY_AGG(block.id))[1],
+          (ARRAY_AGG(block.blockNo))[1],
+          ARRAY_AGG(applyTx.txId)
+        FROM marlowe.applyTx
+        JOIN marlowe.block
+          ON block.id = applyTx.blockId
+        JOIN params USING (createTxId, createTxIx)
+        WHERE block.rollbackToBlock IS NULL
+          AND block.slotNo > params.afterSlot
+        GROUP BY block.slotNo
+        ORDER BY block.slotNo
+        LIMIT 1
+      )
+    , nextWithdrawalTxIds (slotNo, blockHeaderHash, blockNo, txIds) AS
+      ( SELECT
+          block.slotNo,
+          (ARRAY_AGG(block.id))[1],
+          (ARRAY_AGG(block.blockNo))[1],
+          ARRAY_AGG(withdrawalTxIn.txId)
+        FROM marlowe.withdrawalTxIn
+        JOIN marlowe.block
+          ON block.id = withdrawalTxIn.blockId
+        JOIN marlowe.payoutTxOut
+          ON payoutTxOut.txId = withdrawalTxIn.payoutTxId
+          AND payoutTxOut.txIx = withdrawalTxIn.payoutTxIx
+        JOIN marlowe.applyTx
+          ON applyTx.txId = payoutTxOut.txId
+        JOIN params USING (createTxId, createTxIx)
+        WHERE block.rollbackToBlock IS NULL
+          AND block.slotNo > params.afterSlot
+        GROUP BY block.slotNo
+        ORDER BY block.slotNo
+        LIMIT 1
+      )
+    SELECT
+      COALESCE (nextApplyTxIds.slotNo, nextWithdrawalTxIds.slotNo) :: bigint,
+      COALESCE (nextApplyTxIds.blockHeaderHash, nextWithdrawalTxIds.blockHeaderHash) :: bytea,
+      COALESCE (nextApplyTxIds.blockNo, nextWithdrawalTxIds.blockNo) :: bigint,
+      COALESCE (nextApplyTxIds.txIds, '{}') :: bytea[],
+      COALESCE (nextWithdrawalTxIds.txIds, '{}') :: bytea[]
+    FROM nextApplyTxIds
+    FULL OUTER JOIN nextWithdrawalTxIds
+    ON nextApplyTxIds.slotNo = nextWithdrawalTxIds.slotNo
+    ORDER BY COALESCE (nextApplyTxIds.slotNo, nextWithdrawalTxIds.slotNo)
```

Comment:
Would this read more cleanly as a `UNION` instead of a `FULL OUTER JOIN` with all of the `COASLESCE`s? I guess the point is that applications are preferred over withdrawals if both occur in the same slot. In principle, one would like to look at the position of the transaction in the block and order by that, but we don't have such data.

Also, it would be interesting to measure the performance of this on the example contract that involved ~400 transactions.

Reply:
It would not be correct if it was a `UNION` - the purpose of this is to join the two sub queries by block ID, and take the minimum block which has apply transactions, withdraw transactions, or both.